src/library_webgpu.js

```diff
@@ -1842,70 +1938,134 @@ var LibraryWebGPU = {
+    }
+
+    navigator["gpu"]["requestAdapter"](opts)["then"](function(adapter) {
+      if (adapter) {
+        var adapterId = WebGPU.mgrAdapter.create(adapter);
+        {{{ makeDynCall('viiii', 'callback') }}}({{{ gpu.RequestAdapterStatus.Success }}}, adapterId, 0, userdata);
+      } else {
+        var messagePtr = allocateUTF8('WebGPU not available on this system (requestAdapter returned null)');
+        {{{ makeDynCall('viiii', 'callback') }}}({{{ gpu.RequestAdapterStatus.Unavailable }}}, 0, messagePtr, userdata);
+        _free(messagePtr);
+      }
+    }, function(ex) {
+      var messagePtr = allocateUTF8(ex.message);
+      {{{ makeDynCall('viiii', 'callback') }}}({{{ gpu.RequestAdapterStatus.Error }}}, 0, messagePtr, userdata);
+      _free(messagePtr);
+    });
   },
 
   // WGPUAdapter
 
-  wgpuAdapterGetProperties: function() {
-    abort('unimplemented (TODO)');
+  wgpuAdapterGetProperties: function(adapterId, properties) {
+    {{{ gpu.makeCheckDescriptor('properties') }}}
+    {{{ makeSetValue('properties', C_STRUCTS.WGPUAdapterProperties.vendorID, '0', 'i32') }}};
+    {{{ makeSetValue('properties', C_STRUCTS.WGPUAdapterProperties.deviceID, '0', 'i32') }}};
+    {{{ makeSetValue('properties', C_STRUCTS.WGPUAdapterProperties.name, '0', 'i32') }}};
+    {{{ makeSetValue('properties', C_STRUCTS.WGPUAdapterProperties.driverDescription, '0', 'i32') }}};
+    {{{ makeSetValue('properties', C_STRUCTS.WGPUAdapterProperties.adapterType, gpu.AdapterType.Unknown, 'i32') }}};
+    {{{ makeSetValue('properties', C_STRUCTS.WGPUAdapterProperties.backendType, gpu.BackendType.WebGPU, 'i32') }}};
   },
-  wgpuAdapterRequestDevice: function() {
-    abort('unimplemented (TODO)');
+#if MINIMAL_RUNTIME
+  wgpuAdapterRequestDevice__deps: ['$allocateUTF8'],
+#endif
+  wgpuAdapterRequestDevice: function(adapterId, descriptor, callback, userdata) {
+    var adapter = WebGPU.mgrAdapter.get(adapterId);
+
+    var desc = {
+    };
+    if (descriptor) {
+      {{{ gpu.makeCheckDescriptor('descriptor') }}}
+      // TODO: Add support for requiredFeatures and requiredLimits
+    }
+
+    adapter["requestDevice"](desc)["then"](function(device) {
+      var deviceId = WebGPU.mgrDevice.create(device);
+      {{{ makeDynCall('viiii', 'callback') }}}({{{ gpu.RequestDeviceStatus.Success }}}, deviceId, 0, userdata);
+    }, function(ex) {
+      var messagePtr = allocateUTF8(ex.message);
+      {{{ makeDynCall('viiii', 'callback') }}}({{{ gpu.RequestDeviceStatus.Error }}}, 0, messagePtr, userdata);
+      _free(messagePtr);
+    });
   },
 
   // WGPUSurface
 
-  wgpuSurfaceGetPreferredFormat: function(surfaceId, adapterId, callback, userdata) {
-    abort('unimplemented (TODO)');
+  wgpuSurfaceGetPreferredFormat: function(surfaceId, adapterId) {
+    var context = WebGPU.mgrSurface.get(surfaceId);
+    var adapter = WebGPU.mgrAdapter.get(adapterId);
+    const format = context["getPreferredFormat"](adapter);
+    return WebGPU.PreferredFormat[format];
```

Comment:
Why don't we just access TextureFormat here? PreferredFormat is not its own type

Reply:
TextureFormat only has number->string mappings, but since this is an api-output it needs a string->number mapping.

Reply:
Ah okay; makes sense - thanks!